core/cloudflow-spark-testkit/src/main/scala/cloudflow/spark/testkit/SparkStreamletTestkit.scala

```diff
@@ -256,3 +248,44 @@ case class SparkOutletTap[T: Encoder](
   // get results from memory sink
   def asCollection(session: SparkSession): Seq[T] = session.sql(s"select * from $queryName").as[T].collect()
 }
+case class ExecutionReport(totalRows: Long, totalQueries: Int, failures: Seq[String])
+
+class QueryExecutionMonitor()(implicit ec: ExecutionContext) extends StreamingQueryListener {
+  @volatile var status: Map[UUID, QueryState] = Map()
+  @volatile var dataRows: Map[UUID, Long]     = Map()
+
+  sealed trait QueryState
+  case object Started                              extends QueryState
+  case class Terminated(exception: Option[String]) extends QueryState
+
+  def executionReport: ExecutionReport = {
+    val exceptions = status.values.collect { case Terminated(Some(reason)) => reason }
+    ExecutionReport(dataRows.values.sum, dataRows.keys.size, exceptions.toSeq)
+  }
+
+  def waitForData(): Future[Long] = {
+    def hasData   = dataRows.nonEmpty && dataRows.forall { case (_, rows) => rows > 0 }
+    def totalRows = dataRows.values.sum
+    Future {
+      while (!hasData) {
```

Comment:
Why do we wait only until the first bit of data comes in (every dataRow has more than 0 rows?)? Do I understand that correctly? I guess I have the same question as Debasish